ec/internal/planmodifiers/use_state_unless_template_changed.go

```diff
@@ -20,48 +20,144 @@ package planmodifiers
+//  1. The attribute is not nullable (`isNullable = false`) and the topology's state is nil
+//  2. The deployment template attribute has changed
+//  3. `migrate_to_latest_hardware` is set to `true` and there is a migration available to be performed
+func UseStateForUnknownUnlessMigrationIsRequired(resourceKind string, isNullable bool) useStateForUnknownUnlessMigrationIsRequired {
+	return useStateForUnknownUnlessMigrationIsRequired{resourceKind: resourceKind, isNullable: isNullable}
 }
 
-type useStateForUnknownUnlessTemplateChanged struct{}
+type useStateForUnknownUnlessMigrationIsRequired struct {
+	resourceKind string
+	isNullable   bool
+}
+
+type PlanModifierResponse interface {
+	planmodifier.StringResponse | planmodifier.Int64Response
+}
 
-func (m useStateForUnknownUnlessTemplateChanged) PlanModifyString(ctx context.Context, req planmodifier.StringRequest, resp *planmodifier.StringResponse) {
-	if !resp.PlanValue.IsUnknown() {
-		return
+func (m useStateForUnknownUnlessMigrationIsRequired) PlanModifyString(ctx context.Context, req planmodifier.StringRequest, resp *planmodifier.StringResponse) {
+	useState, diags := m.UseState(ctx, req.ConfigValue, req.Plan, req.State, resp.PlanValue, req.StateValue)
+	resp.Diagnostics.Append(diags...)
+	if useState {
+		resp.PlanValue = req.StateValue
 	}
+}
 
-	// if the config is the unknown value, use the unknown value otherwise, interpolation gets messed up
-	if req.ConfigValue.IsUnknown() {
-		return
+func (m useStateForUnknownUnlessMigrationIsRequired) PlanModifyInt64(ctx context.Context, req planmodifier.Int64Request, resp *planmodifier.Int64Response) {
+	useState, diags := m.UseState(ctx, req.ConfigValue, req.Plan, req.State, resp.PlanValue, req.StateValue)
+	resp.Diagnostics.Append(diags...)
+	if useState {
+		resp.PlanValue = req.StateValue
 	}
+}
+
+func (m useStateForUnknownUnlessMigrationIsRequired) UseState(ctx context.Context, configValue attr.Value, plan tfsdk.Plan, state tfsdk.State, planValue attr.Value, stateValue attr.Value) (bool, diag.Diagnostics) {
+	var diags diag.Diagnostics
 
-	if req.StateValue.IsNull() {
-		return
+	if stateValue.IsNull() && !m.isNullable {
+		return false, nil
 	}
 
-	templateChanged, diags := AttributeChanged(ctx, path.Root("deployment_template_id"), req.Plan, req.State)
-	resp.Diagnostics.Append(diags...)
+	if !planValue.IsUnknown() {
+		return false, nil
+	}
+
+	// if the config is the unknown value, use the unknown value otherwise, interpolation gets messed up
+	if configValue.IsUnknown() {
+		return false, nil
+	}
+
+	templateChanged, d := AttributeChanged(ctx, path.Root("deployment_template_id"), plan, state)
+	diags.Append(d...)
+
+	var migrateToLatestHw bool
+	plan.GetAttribute(ctx, path.Root("migrate_to_latest_hardware"), &migrateToLatestHw)
+
+	isMigrationAvailable, d := CheckAvailableMigration(ctx, plan, state, path.Root(m.resourceKind))
```

Comment:
nit-pick: if `migrate_to_latest_hardware` isn't set, there is no need to check that migration is available

Reply:
Great point! 😉 added in https://github.com/elastic/terraform-provider-ec/pull/755/commits/a69026c9d9ed2acf9c0f3872779d6be75fb05bb6.

Reply:
Thanks for the tip 🙏 I did a little refactor in https://github.com/elastic/terraform-provider-ec/pull/755/commits/7860b3b95a00c994967cd8a0e71b67508c262a15, based on this strategy.